{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import itertools\n",
    "from tqdm import tqdm\n",
    "\n",
    "import numpy as np\n",
    "import scipy.stats as sps\n",
    "\n",
    "\n",
    "def get_kmers(dna, k):\n",
    "    return [np.array(list(dna[i : i + k]))\n",
    "            for i in range(len(dna) - k)]\n",
    "\n",
    "\n",
    "def get_base_counts(dna):\n",
    "    bases = ['A', 'T', 'G', 'C']\n",
    "    return np.array([\n",
    "        np.sum(dna == base) \n",
    "        for base in bases\n",
    "    ])\n",
    "\n",
    "\n",
    "def get_base_pseudocounts(dna):\n",
    "    return get_base_counts(dna) + 1\n",
    "\n",
    "\n",
    "base_to_num = {\n",
    "    'A' : 0, 'T' : 1,\n",
    "    'G' : 2, 'C' : 3\n",
    "}\n",
    "\n",
    "num_to_base = {\n",
    "    val : key\n",
    "    for key, val in base_to_num.items()\n",
    "}\n",
    "\n",
    "\n",
    "def get_profile(motifs):\n",
    "    motif_mx = np.vstack(motifs)\n",
    "    t, k = motif_mx.shape\n",
    "    counts = np.column_stack([\n",
    "        get_base_pseudocounts(col)\n",
    "        for col in motif_mx.T\n",
    "    ])\n",
    "    assert counts.shape == (4, k)\n",
    "    return counts / counts.sum(axis=0)\n",
    "\n",
    "\n",
    "def eval_kmer_proba(kmer, profile):\n",
    "    return np.prod([\n",
    "        profile[base_to_num[kmer[i]], i]    \n",
    "        for i in range(len(kmer))\n",
    "    ])\n",
    "\n",
    "\n",
    "\n",
    "def find_most_probable_motif(dna, k, profile):\n",
    "    kmers_with_probas = np.vstack([\n",
    "        (kmer, eval_kmer_proba(kmer, profile))\n",
    "        for kmer in get_kmers(dna, k)\n",
    "    ])\n",
    "    top_row = np.argmax(kmers_with_probas[:, 1])\n",
    "    return kmers_with_probas[top_row, 0]\n",
    "\n",
    "\n",
    "def get_consensus(profile):\n",
    "    return np.array([\n",
    "        num_to_base[np.argmax(col)]\n",
    "        for col in profile.T\n",
    "    ])\n",
    "\n",
    "\n",
    "def get_score(motifs):\n",
    "    motif_mx = np.vstack(motifs)\n",
    "    profile = get_profile(motif_mx)\n",
    "    consensus = get_consensus(profile)\n",
    "    return np.sum(\n",
    "        (motif != consensus).sum()\n",
    "        for motif in motif_mx\n",
    "    )\n",
    "\n",
    "\n",
    "def randomized_motif_search(reads, k, t):\n",
    "    n = len(reads[0])\n",
    "    motifs = [\n",
    "        np.array(list(reads[i][start : start + k]))\n",
    "        for i, start in enumerate(np.random.randint(low=0, high=n-k, size=t))\n",
    "    ]\n",
    "    best_motifs = motifs\n",
    "    while True:\n",
    "        profile = get_profile(motifs)\n",
    "        motifs = [\n",
    "            find_most_probable_motif(dna, k, profile)\n",
    "            for dna in reads\n",
    "        ]\n",
    "        if get_score(motifs) < get_score(best_motifs):\n",
    "            best_motifs = motifs\n",
    "        else: \n",
    "            return best_motifs\n",
    "\n",
    "\n",
    "def sample_from_profile(dna, profile):\n",
    "    k = profile.shape[1]\n",
    "    kmers_with_probas = np.vstack([\n",
    "        (kmer, eval_kmer_proba(kmer, profile))\n",
    "        for kmer in get_kmers(dna, k)\n",
    "    ])\n",
    "    kmers = kmers_with_probas[:, 0]\n",
    "    probas = kmers_with_probas[:, 1]\n",
    "    probas /= probas.sum()\n",
    "    kmer_dist = sps.rv_discrete(values=(np.arange(len(kmers)), probas))\n",
    "    return kmers[kmer_dist.rvs()]\n",
    "    \n",
    "\n",
    "def gibbs_sampler_motif_search(reads, k, t, n_restarts=100):\n",
    "    n = len(reads[0])\n",
    "    motifs = [\n",
    "        np.array(list(reads[i][start : start + k]))\n",
    "        for i, start in enumerate(np.random.randint(low=0, high=n-k, size=t))\n",
    "    ]\n",
    "    best_motifs = motifs\n",
    "    for _ in range(n_restarts):\n",
    "        i = np.random.randint(t)\n",
    "        profile = get_profile(motifs[:i] + motifs[i+1:])\n",
    "        missing_motif = sample_from_profile(reads[i], profile)\n",
    "        motifs[i] = missing_motif\n",
    "        if get_score(motifs) < get_score(best_motifs):\n",
    "            best_motifs = motifs\n",
    "    return best_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:78: DeprecationWarning: Calling np.sum(generator) is deprecated, and in the future will give a different result. Use np.sum(np.fromiter(generator)) or the python sum builtin instead.\n",
      "gibbs sampler iterations passed:   0%|          | 0/20 [00:00<?, ?it/s]/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/numpy/core/_asarray.py:136: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
      "  return array(a, dtype, copy=False, order=order, subok=True)\n",
      "gibbs sampler iterations passed: 100%|██████████| 20/20 [14:49<00:00, 44.46s/it]\n"
     ]
    }
   ],
   "source": [
    "lines = [line.strip() for line in list(open(\"gibbs.txt\", \"r\"))]\n",
    "#target_motifs = [np.array(list(line.strip())) for line in list(open(\"target.txt\", \"r\"))]\n",
    "k, t, n_restarts = (int(num) for num in lines[0].split())\n",
    "reads = lines[1:]\n",
    "MC_ITERS = 20\n",
    "best_motifs = [ \n",
    "    np.array(list(dna[:k]))\n",
    "    for dna in reads\n",
    "]  \n",
    "best_score = get_score(best_motifs)\n",
    "outfile = open(\"answers.txt\", \"w\")\n",
    "for _ in tqdm(range(MC_ITERS), desc=\"gibbs sampler iterations passed\"):\n",
    "    candidate_motifs = gibbs_sampler_motif_search(reads, k, t, n_restarts)\n",
    "    candidate_score = get_score(candidate_motifs)\n",
    "    if candidate_score < best_score:\n",
    "        best_score = candidate_score\n",
    "        best_motifs = candidate_motifs\n",
    "        #print(\"CURRENT\", best_score)\n",
    "        #print(\"TARGET: \", get_score(target_motifs))\n",
    "        outfile.seek(0)        \n",
    "        answers = '\\n'.join([''.join(motif) for motif in best_motifs])\n",
    "        outfile.write(answers)\n",
    "outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A', 'C', 'G', 'T', 'C', 'C', 'A', 'C', 'C', 'G', 'G', 'C', 'G',\n",
       "       'T', 'C'], dtype='<U1')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.row_stack(best_motifs)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
